### ISSUE TITLE
Wormhole and API server instances replicas

### ISSUE THREAD

**Issue by WinterNis** [sapcc/kubernikus#300]
Hi,

## What I leaned by reading the code

- Wormhole clients (with guttle) are in charge of creating a tunnel between api server instance(s) and nodes, allowing the apiserver to contact pod and services through their PodCidr IP and ServiceCidr IP.
- Wormhole clients are deployed on klusters nodes, and are using the kluster url to connect to the wormhole server.
- kluster url resolve on a load-balancer (an ingress). Therefore, when wormhole clients initiate the connection to the server, it connects to one of the pod of the service resolved by the ingress, and create a tunnel for this one apiserver pod. Tunnel for other apiserver pods will not be created.
- With the actual implementation of the wormhole, it is not possible to deploy more than one api server pod in a kluster.
- Request to serviceCidr from API Server are going through a random tunnel.

First of all, can you, please, confirm what I wrote?

## What I think about it

If that is the case, you should agree that using only one instance of apiserver is unpractical.
With several instances, we could lower the unavailability time in case of node failure, or during rolling upgrades.

In my opinion, the main issue with the actual design is that the connection is instanciated from the node, which is unpractical since it is very difficult for the node to know about all apiserver instances.
On the opposite, it is really easy for apiserver instances to know about nodes. 

## What I suggest

I have implemented a proof-of-concept of a solution to this problem, see #299.
It seems that Guttle server and client has been designed to work in a symetric way.
Meaning that, like the guttle server, the guttle client can listen for incoming traffic and redirect it into the tunnel, sending headers with the original destination IP and port. Like the guttle client, the guttle server can take a proxy function in parameter and use this function to dial connection on its tunnel side. 

TLDR:
I inverted the client/server concept and instanciation. Now one Wormhole server is deployed on each node, and multiple wormhole clients are instanciated in the apiserver pod (one for each node)   

Here are quick schemas of how things work at present, and how things work in what I implemented:

Actual wormhole design : 
![original wormhole](https://user-images.githubusercontent.com/12795509/43405034-3634d938-9419-11e8-9659-79450662c4a3.png)

Actual wormhole details : 
![original wormhole details](https://user-images.githubusercontent.com/12795509/43405183-8e38ffba-9419-11e8-95bc-daa788f7154e.png)

Wormhole design proposition : 
![ovh-wormhole](https://user-images.githubusercontent.com/12795509/43405306-d8c210a8-9419-11e8-8df3-f21ccca3f0d8.png)

Node access to API server service is not represented here. 

Wormhole proposition details : 
![ovh project details](https://user-images.githubusercontent.com/12795509/43405322-df9036b2-9419-11e8-9239-acba082d0900.png)

## What next

I identified one issue with this proposition:
- A Guttle client is instantiated for each node (by each api server). WormHole clients can not route serviceCidr IP with Guttle route: Guttle routes concept are configured on server side.
We have to implements a way to randomized the request on nodes. Since Guttle is close to be symetrical, I suggest that we try to make all features available for both client and server. it would suit nicely to the use case. 

Note:
- I did not configure the wormhole to allow node to contact apiserver with its service IP. This could be done using hijjack-address on the node and a route pointing on the api-server load-balancer IP..
- For the sake of easy testing and POC'ing I did not plug back the code and templates responsible for instanciating the wormhole in the api-server and on nodes (helm and ignitions templates)

I did not push further this work, since I wanted to know what you would think of it. Is this relevant to kubernikus? Do you think it deserves to be pushed further?

Thanks

**Comment by BugRoger** [comment]
You reverse engineered all this? Hat off! I'm flabbergasted... 😮

I'll leave this for @databus23 to answer but in general, I think, you analysis of the code is spot on.

**Comment by BugRoger** [comment]
Accidental close. /reopen

**Comment by WinterNis** [comment]
Thanks for the answer ! 

> This is the root why we need the Wormhole in the first place. In our case the nodes are sealed off in a private network - unaccessible from anywhere else. The Apiserver is running outside and can't directly talk to the nodes. Hence the nodes need to open the tunnel.

Humf, I did not know about this. 
On one hand it makes sense, on the other hand it looks a bit unpractical. Out of curiosity, what is the actual reason behind the private network ?
Also, I think that the wormhole is necessary even with public accessible nodes, since apiserver in pods would still not be able to dial pods with PodCidr IPs or services with ServiceCidr IPS, am I right ? 

Also, do you confirm that this matter is the only thing restraining the number of apiserver replicas ? Or did I miss something else ?    

Thanks

**Comment by BugRoger** [comment]
> On one hand it makes sense, on the other hand it looks a bit unpractical. Out of curiosity, what is the actual reason behind the private network ?

Each customer is running their own private network in Openstack. The purpose is isolation and security. They can create a Kubernikus cluster in their network. The apiserver is running in a regional control-plane. The node's are running in the customer's private network. By default the private network and with that the nodes are not accessible from the outside. It is possible to expose individual nodes via floating ips and fence them with security groups, but we didn't want to have this as a requirement. Ideally, you only want to expose specific services.

We're running all customer apiservers in a dedicated Kubernetes cluster. The cluster itself is unaccessible from the outside. The apiservers on the other side are exposed via ingress and SNI.

> Also, I think that the wormhole is necessary even with public accessible nodes, since apiserver in pods would still not be able to dial pods with PodCidr IPs or services with ServiceCidr IPS, am I right ?

Yes, that is right. Whether the apiserver actually does this is another question. I suppose the logs and exec calls are proxied through the Kubelet. 

> Also, do you confirm that this matter is the only thing restraining the number of apiserver replicas ? Or did I miss something else ?

I'm not 100% positive we actually do have a restriction on the number of replica's per Apiserver. I think, we could have multiple apiserver pods including multiple wormhole servers. Since they are hidden behind an ingress and service the connection would just load-balance between the pods. We need @databus23 to comment on this one.

**Comment by WinterNis** [comment]
Using a private network does seem justified and secure. Thanks for the explanation. The floating IP trick is nice. 

> Yes, that is right. Whether the apiserver actually does this is another question. I suppose the logs and exec calls are proxied through the Kubelet.

I'm pretty sure that logs, exec, and port-forward will work out the box, since, as you pointed out, it is proxied through the Kubelet. However, the apiserver proxy feature will not work, because in this case the apiserver makes request directly on PodCidr IPS and ServiceCidr IPS.

> I'm not 100% positive we actually do have a restriction on the number of replica's per Apiserver. I think, we could have multiple apiserver pods including multiple wormhole servers. Since they are hidden behind an ingress and service the connection would just load-balance between the pods. We need @databus23 to comment on this one.

As you say, Node -> Apiserver flow will work with the actual design with multiple API instances. However Apiserver -> Node instanciated flow (apiserver proxy for example) will only work with one api server instance, the one resolved by the ingress/service at wormhole connection time. I also not 100% sure of this, so let's wait for @databus23 to answer/confirm :)
